File: DD_c_plusplus/Examples/IndependentSetInstance/IndependentSetProblem.cpp
#include "IndependentSetProblem.h"
#include <numeric>

IndependentSetProblem::IndependentSetProblem(const IndependentSetInstance& params, bool sort)
        : AbstractProblem(params.initial_state, params.variables)
{
    _dict_node_neighbors = params.dict_node_neighbors;
    check_atributes(params.variables, params.dict_node_neighbors);

    // Identity default (overwritten by sort_variables when sort=true)
    dd_to_original_index.resize(params.variables.size());
    iota(dd_to_original_index.begin(), dd_to_original_index.end(), 0);

    if (sort) {
        apply_variable_order(sort_variables(params.variables));
    }

    int n = static_cast<int>(params.dict_node_neighbors.size());
    node_neighbors.resize(n);
    for (const auto& [key, neighbors] : params.dict_node_neighbors) {
        int var_index = this->name_to_index.at(key);
        node_neighbors[var_index] = dynamic_bitset<>(n);
        for (int neighbor_num : neighbors) {
            int neighbor_index = this->name_to_index.at("x_" + to_string(neighbor_num));
            node_neighbors[var_index].set(neighbor_index);
        }
    }
}

vector<string> IndependentSetProblem::sort_variables(const vector<pair<string, vector<int>>>& variables) {
    vector<string> sorted_names;
    sorted_names.reserve(variables.size());
    for (const auto& [name, domain] : variables)
        sorted_names.push_back(name);
    std::stable_sort(sorted_names.begin(), sorted_names.end(),
        [this](const string& a, const string& b) {
            return _dict_node_neighbors.at(a).size() > _dict_node_neighbors.at(b).size();
        });

    map<string, int> name_to_orig;
    for (int i = 0; i < (int)variables.size(); ++i)
        name_to_orig[variables[i].first] = i;
    dd_to_original_index.clear();
    for (const string& name : sorted_names)
        dd_to_original_index.push_back(name_to_orig.at(name));

    return sorted_names;
}

void IndependentSetProblem::check_atributes(const vector<pair<string, vector<int>>>& variables, const map<string, vector<int>>& dict_node_neighbors) {
    check_same_variables(variables, dict_node_neighbors);
    check_consistent_dictionary_of_neighbors(dict_node_neighbors);
}

void IndependentSetProblem::check_same_variables(const vector<pair<string, vector<int>>>& variables, const map<string, vector<int>>& dict_node_neighbors) {
    assert(keysOfVector(variables) == keysToVector(dict_node_neighbors) && "Variables must be the same between dictionaries");
}

unordered_set<std::string> IndependentSetProblem::keysToVector(const map<string, vector<int>>& dict_node_neighbors) {
    std::unordered_set<std::string> result;
    for (const auto& entry : dict_node_neighbors) {
        result.insert(entry.first);
    }
    return result;
}

unordered_set<std::string> IndependentSetProblem::keysOfVector(const vector<pair<string, vector<int>>>& variables) {
    std::unordered_set<std::string> result;
    for (const auto& entry : variables) {
        result.insert(entry.first);
    }
    return result;
}

void IndependentSetProblem::check_consistent_dictionary_of_neighbors(const map<string, vector<int>>& dict_node_neighbors) {
    for (const auto& entry : dict_node_neighbors) {
        for (int value : entry.second) {
            assert(isNeighborConsistent(entry.first, value, dict_node_neighbors) && "Dictionary of neighbors must be consistent");
        }
    }
}

bool IndependentSetProblem::isNeighborConsistent(const string& key, int value, const map<string, vector<int>>& dict_node_neighbors) {
    auto neighborEntry = dict_node_neighbors.find("x_" + to_string(value));
    return (neighborEntry != dict_node_neighbors.end()) &&
           (find(neighborEntry->second.begin(), neighborEntry->second.end(), stoi(key.substr(2))) != neighborEntry->second.end());
}

bool IndependentSetProblem::transition_function(const dynamic_bitset<>* previous_state, const int variable_index, int variable_value, dynamic_bitset<>* scratch_state) const {
    if (variable_value == 0) {
        *scratch_state = *previous_state;
        if (previous_state->test(variable_index))
            scratch_state->reset(variable_index);
        return true;
    }
    if (!previous_state->test(variable_index))
        return false;
    *scratch_state = *previous_state;
    *scratch_state -= node_neighbors[variable_index];
    scratch_state->reset(variable_index);
    return true;
}

int IndependentSetProblem::get_priority_for_discard_node(const dynamic_bitset<>* state) const {
    return int(state->count());
}

int IndependentSetProblem::get_priority_for_merge_nodes(const int node_id, const dynamic_bitset<>* state) const {
    return int(state->count());
}

dynamic_bitset<>* IndependentSetProblem::merge_operator(const dynamic_bitset<>* state_one, const dynamic_bitset<>* state_two) const {
    dynamic_bitset<>* merged_state = new dynamic_bitset<>(*state_one | *state_two);
    return merged_state;
}

string IndependentSetProblem::get_state_as_string(const dynamic_bitset<>* state) const {
    string state_string;
    boost::to_string(*state, state_string);
    return state_string;
}

dynamic_bitset<>* IndependentSetProblem::get_state_copy(const dynamic_bitset<>* state) const {
    dynamic_bitset<>* copy_bitset = new dynamic_bitset<>(*state);
    return copy_bitset;
}

dynamic_bitset<>* IndependentSetProblem::convert_to_bitset(const vector<int>& vec, int size) {

    dynamic_bitset<>* bs = new dynamic_bitset<>(size);

    for (int value : vec) {
        if (value > 0) {
            bs->set(value - 1);
        }
    }
    return bs;
}

dynamic_bitset<>* IndependentSetProblem::get_final_state() {
    vector<int> final_state(node_neighbors.size(), 0);
    return convert_to_bitset(final_state, node_neighbors.size());
}

File: DD_python/Examples/IndependentSetInstance/IndependentSetProblem.py
import os
import sys
from bitarray import bitarray

current_dir = os.path.dirname(os.path.abspath(__file__))
parent_dir = os.path.abspath(os.path.join(current_dir, os.pardir))
sys.path.append(parent_dir)

from SourceCode.Problems.AbstractProblemClass import AbstractProblem

class IndependentSetProblem(AbstractProblem):

    def __init__(self, params: 'IndependentSetStructure', sort: bool = True) -> None:
        '''
        Initialize the Independent Set Problem instance.

        Parameters:
        - params (IndependentSetStructure): Instance of the independent set problem, containing:
            * initial_state: initial bitset state.
            * variables (list[tuple[str, list[int]]]): variable names and their domains.
            * dict_node_neighbors (dict[str, list[int]]): adjacency list per node.
        - sort (bool): whether to sort variables by heuristic order.
        '''
        self._dict_node_neighbors = params.dict_node_neighbors
        super().__init__(params.initial_state, params.variables, sort)

        n = len(params.dict_node_neighbors)
        self.node_neighbors: list[bitarray] = [bitarray(n) for _ in range(n)]
        for bs in self.node_neighbors:
            bs.setall(False)

        for key, neighbors in params.dict_node_neighbors.items():
            var_index = self.name_to_index[key]
            for neighbor_num in neighbors:
                self.node_neighbors[var_index][self.name_to_index[f"x_{neighbor_num}"]] = True

        self.check_atributes(params.variables, params.dict_node_neighbors)


    def sort_variables(self, variables: list[tuple['Variable', 'Domain']]) -> list[str]:
        sorted_names = sorted(
            [name for name, _ in variables],
            key=lambda name: len(self._dict_node_neighbors[name]),
            reverse=True
        )
        name_to_original_index = {name: i for i, (name, _) in enumerate(variables)}
        self.dd_to_original_index = [name_to_original_index[name] for name in sorted_names]
        return sorted_names

    def check_atributes(self, variables: list[tuple['Variable', 'Domain']], dict_node_neighbors: dict[str, list[int]]) -> None:
        ''' 
        Check the attributes of the problem.

        Parameters:
        - variables (list[tuple['Variable', 'Domain']]): List of tuples with the variables and their domains.
        '''
        self.check_same_variables(variables,dict_node_neighbors)
        self.check_neighbors_must_be_integers(dict_node_neighbors)
        self.check_consistent_dictionary_of_neighbors(dict_node_neighbors)

    def check_same_variables(self, variables: list[tuple['Variable', 'Domain']], dict_node_neighbors: dict[str, list[int]]) -> None:
        '''
        Check if the variables are the same between the dictionary of neighbors and the list of variables.

        Parameters:
        - variables (list[tuple['Variable', 'Domain']]): List of tuples with the variables and their domains.
        '''
        assert dict(variables).keys() == dict_node_neighbors.keys(), "Variables must be the same between dictionaries"
    
    def check_neighbors_must_be_integers(self, dict_node_neighbors: dict[str, list[int]]) -> None:
        ''' 
        Check if the neighbors are integers.
        '''
        for key in dict_node_neighbors.keys():
            for value in dict_node_neighbors.get(key, []):
                assert isinstance(value, int), "Values must be integers"
    
    def check_consistent_dictionary_of_neighbors(self, dict_node_neighbors: dict[str, list[int]]) -> None:
        '''
        Check if the dictionary of neighbors is consistent.
        '''
        for key in dict_node_neighbors.keys():
            for value in dict_node_neighbors.get(key, []):
                assert int(key[2:]) in dict_node_neighbors.get("x_"+str(value), []), "Dictionary of neighbors must be consistent"

    def transition_function(self, previous_state: 'State', variable_index: int, variable_value: int, scratch_state: list) -> bool:
        if variable_value == 0:
            scratch_state[0][:] = previous_state
            if previous_state[variable_index]:
                scratch_state[0][variable_index] = False
            return True
        if not previous_state[variable_index]:
            return False
        scratch_state[0][:] = previous_state
        scratch_state[0] &= ~self.node_neighbors[variable_index]
        scratch_state[0][variable_index] = False
        return True

    def get_priority_for_discard_node(self, state: 'State') -> int:
        '''
        Get the priority for discard a node.

        Parameters:
        - state ('State'): State of the problem.

        Returns:
        - int: Priority for discard the node.
        '''
        return state.count(True)
    
    def get_priority_for_merge_nodes(self, node_id: int, state: 'State') -> int:
        '''
        Get the priority for merge two nodes.

        Parameters:
        - id (int): Id of the node.
        - state ('State'): State of the problem.

        Returns:
        - int: Priority for merge the nodes.
        '''
        return state.count(True)

    def merge_operator(self, state_one: 'State', state_two: 'State') -> 'State':
        '''
        Merge two states.

        Parameters:
        - state_one ('State'): First state.
        - state_two ('State'): Second state.

        Returns:
        - 'State': Merged state.
        '''
        return state_one | state_two
    
    def get_state_as_string(self, state: 'State') -> str:
        '''
        Get the state as a string.

        Parameters:
        - state ('State'): State of a node.

        Returns:
        - str: State as a string.
        '''
        return state.to01()
    
    def get_state_copy(self, state: 'State') -> 'State':
        '''
        Copy of a state.

        Parameters:
        - state(State): The state to copy. *READ IMPORTANT

        Returns:
        - State: A copy of the state.
        '''
        return state.copy()
    
    def get_final_state(self) -> 'State':
        '''
        Final state of the problem.

        Returns:
        - State: The final state of the problem.
        '''
        final_state: list[int] = [0] * len(self.node_neighbors)
        return self.convert_to_bitset(final_state, len(self.node_neighbors))

    def convert_to_bitset(self, vec, size):
        '''
        Convert a list to a bitset.
        '''
        bs: bitarray = bitarray(size)
        bs.setall(False)
        for value in vec:
            if value > 0:
                bs[value - 1] = True
        return bs
